src/Routers/FunctionsRouter.js

```diff
@@ -106,69 +115,66 @@ export class FunctionsRouter extends PromiseRouter {
-        const userString = (req.auth && req.auth.user) ? req.auth.user.id : undefined;
-        const cleanInput = logger.truncateLogMessage(JSON.stringify(params));
-        var response = FunctionsRouter.createResponseObject((result) => {
-          try {
-            const cleanResult = logger.truncateLogMessage(JSON.stringify(result.response.result));
-            logger.info(
-              `Ran cloud function ${functionName} for user ${userString} with:\n  Input: ${cleanInput }\n  Result: ${cleanResult }`,
-              {
-                functionName,
-                params,
-                user: userString,
-              }
-            );
-            resolve(result);
-          } catch (e) {
-            reject(e);
-          }
-        }, (error) => {
-          try {
-            logger.error(
-              `Failed running cloud function ${functionName} for user ${userString} with:\n  Input: ${cleanInput}\n  Error: ` + JSON.stringify(error),
-              {
-                functionName,
-                error,
-                params,
-                user: userString
-              }
-            );
-            reject(error);
-          } catch (e) {
-            reject(e);
-          }
-        });
-        // Force the keys before the function calls.
-        Parse.applicationId = req.config.applicationId;
-        Parse.javascriptKey = req.config.javascriptKey;
-        Parse.masterKey = req.config.masterKey;
-        theFunction(request, response);
+    return new Promise(function (resolve, reject) {
+      const userString = (req.auth && req.auth.user) ? req.auth.user.id : undefined;
+      const cleanInput = logger.truncateLogMessage(JSON.stringify(params));
+      const { success, error, message } = FunctionsRouter.createResponseObject((result) => {
+        try {
+          const cleanResult = logger.truncateLogMessage(JSON.stringify(result.response.result));
+          logger.info(
+            `Ran cloud function ${functionName} for user ${userString} with:\n  Input: ${cleanInput }\n  Result: ${cleanResult }`,
+            {
+              functionName,
+              params,
+              user: userString,
+            }
+          );
+          resolve(result);
+        } catch (e) {
+          reject(e);
+        }
+      }, (error) => {
+        try {
+          logger.error(
+            `Failed running cloud function ${functionName} for user ${userString} with:\n  Input: ${cleanInput}\n  Error: ` + JSON.stringify(error),
+            {
+              functionName,
+              error,
+              params,
+              user: userString
+            }
+          );
+          reject(error);
+        } catch (e) {
+          reject(e);
+        }
       });
-    } else {
-      throw new Parse.Error(Parse.Error.SCRIPT_FAILED, `Invalid function: "${functionName}"`);
-    }
+      return Promise.resolve().then(() => {
+        return theFunction(request, { message });
+      }).then(success, error);
+    });
```

Comment:
why the extra `Promise.resolve().then() again? why not 

`return theFunction(request, { message }).then(success).catch(error)`?

Reply:
nope :) that's not what's intended, in case of error in the susccess, we don,t want to use the current error handler but the next one.